core/src/main/scala/Metrics.scala

```diff
@@ -156,13 +159,17 @@ object Metrics {
   /*
    * Perform a timed operation with configurable handlers.
    */
-  def timer(before: Task[Unit], onComplete: Kleisli[Task, Double, Unit], onFail: Kleisli[Task, Throwable, Unit], onSuccess: Task[Unit]) = new (Task ~> Task) {
-    def apply[A](op: Task[A]) : Task[A] = (for {
-      _ <- before
-      start <- Task.delay(System.nanoTime())
-      a <- op
-      elapsed = (System.nanoTime - start) / 1.0e9
-      _ <- onComplete.run(elapsed)
-    } yield a).onFinish(_.cata(onFail.run, onSuccess))
+  def timer(before: IO[Unit], onComplete: Kleisli[IO, Double, Unit], onFail: Kleisli[IO, Throwable, Unit], onSuccess: IO[Unit]) = new (IO ~> IO) {
+    def apply[A](op: IO[A]) : IO[A] = {
+      val io = for {
+        _       <- before
+        start   <- IO(System.nanoTime())
+        a       <- op
+        elapsed = (System.nanoTime - start) / 1.0e9
+        _       <- onComplete.run(elapsed)
+      } yield a
+
+      (io <* onSuccess).onError { case err => onFail(err) }
```

Comment:
Where does `onError` come from?  Why is this not a `handleWith`?

Reply:
`onError` is from `ApplicativeError`: https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/ApplicativeError.scala#L99-L128

I believe the use here is primarily convenience, we just want to run a handler and then give back the original result. Could do it with `handleErrorWith` with some dancing about. I also wonder if it might be better to just do something like:

```scala
io.attempt.flatMap {
  case Left(e) => onFail(e) *> IO.raiseError(e)
  case Right(a) => onSuccess(a) *> IO.pure(a)
}
```

As far as I know this is what the Scalaz equivalent of `onFinish` was doing: https://github.com/scalaz/scalaz/blob/series/7.3.x/concurrent/src/main/scala/scalaz/concurrent/Task.scala#L50-L61

I'm also now wondering why it isn't just implemented the way I did above..